XMonad/WorkspaceLayout/Grid.hs

```diff
@@ -0,0 +1,326 @@
+  []     -> Mapping mempty
+  (x0:_) -> Mapping $ xs & foldMap (\x -> Map.singleton x (doFormat @ftd x0 name))
+
+
+-- | Glue together a column of workspaces
+column :: forall ftd. IsFormatted ftd => Dims -> Int -> String -> Mapping ftd
+column (Dims { height }) x name =
+  Mapping . fold $ do
+    y <- [0 .. height - 1]
+    let topLeft = XY x 0
+    pure $ Map.singleton (XY x y) (doFormat @ftd topLeft name)
+
+
+
+
+data Coord = XY { x :: Int, y :: Int }
+  deriving (Show, Ord, Eq, Generic)
+
+data Wrapping = Wrapping
+  { wrapX :: Bool
+  , wrapY :: Bool
+  }
+  deriving (Show, Generic)
+
+data State = State
+  { mapping  :: SomeMapping
+      -- ^ Coordinate -> WorkspaceId mapping
+  , labelf   :: Coord -> Maybe String
+      -- ^ Labels
+  , wrapping :: Wrapping
+      -- ^ Wrapping mode
+  , coord    :: Coord
+      -- ^ Current coordinate
+  } deriving (Generic)
+
+instance St.OneState State where
+  type Mod State = State -> State
+  merge ma s = pure (ma s)
+  defaultState = State
+    { mapping = SomeMapping (grid $ Dims 5 5)
+    , wrapping = Wrapping False False
+    , coord = XY 0 0
+    , labelf = const Nothing
+    }
+
+-- Wrap a coordinate around the x/y axes according to the configured wrapping mode
+wrap :: State -> (Coord -> Coord)
+wrap state =
+
+  appEndo . fromMaybe (Endo id) $ do
+    xRange <- range x (mapping state)
+    yRange <- range y (mapping state)
+
+    pure $ guard (state & wrapping & wrapX) (Endo $ \c -> c { x = affineMod xRange (x c) })
+        <> guard (state & wrapping & wrapY) (Endo $ \c -> c { y = affineMod yRange (y c) })
+
+  where
+
+  guard :: Monoid m => Bool -> m -> m
+  guard = \case { True -> id; False -> const mempty }
+
+
+
+-- |
+--
+-- The call @move f@ replaces the current coordinate to @f currentCoord@
+--
+-- If @f currentCoord@ is out-of-bounds, do nothing
+--
+-- Use this to move around workspaces
+move :: (Coord -> Coord) -> X ()
+move = update $ \coord wid -> do
+  St.modify $ \st -> st { coord = coord }
+  windows (greedyView wid)
+
+
+-- |
+--
+-- The call @swap f@ moves the selected window to @f currentCoord@
+--
+-- If @f currentCoord@ is out-of-bounds, do nothing
+swap :: (Coord -> Coord) -> X ()
+swap = update $ \_ wid -> windows (shift wid)
+
+update :: (Coord -> WorkspaceId -> X ()) -> (Coord -> Coord) -> X ()
+update act f = do
+  state@State { coord, mapping } <- St.get
+  let coord' = coord & f & wrap state
+  case Map.lookup coord' (getTheMap mapping) of
+    Nothing -> pure ()
+    Just wid -> do
+     act coord' wid
+
+
+data Init = Init
+  { initMapping  :: SomeMapping
+      -- ^ Physical layout
+  , initWrapping :: Wrapping
+      -- ^ Wrapping mode
+  , initLabelf   :: Coord -> Maybe String
+      -- ^
+      -- Displayed to the left of the axis workspaces
+      -- When Nothing is returned, uses a fallback
+  }
+
+-- | Hook the grid layout into XMonad
+hook :: Init -> XConfig l -> XConfig l
+hook Init { initMapping, initWrapping, initLabelf } =
```

Comment:
This should be named differently

Reply:
Fine by me. Suggestions?

Reply:
People commonly import xmonad-contrib modules unqualified, so the user-facing functions should be named in a way that works without the module name. In this case, `gridWorkspaceLayout` is probably best.

You'd need to unexport or rename all the other functions too… Perhaps it'd be okay to include a Usage section (most if not all other modules do have such a section, so you should have one here too) that instructs users to import it qualified, and then the names could stay.

Reply:
I'm against using qualified imports here, it's really not something that the modules do.

Reply:
Is the consensus to switch to unqualified-import-friendly names? Exporting something like

```haskell
gridGrid :: GridDims -> GridMapping 'GridFormatted
```

seems very clunky to me. And I don't think it's the responsibility of the library author to manage the user's imports.

That being said, I'll follow consensus